docs/development/errors-diagnostic.md

```diff
@@ -0,0 +1,103 @@
+
+System-defined errors caused by the operating system cover most cases, such as `EBUSY`, `ENOSYS` or `EFAULT`. While several errors resulted by the ONNC compiler, for example, `Success`, `NotStartedYet`, or `UnknownError`, are also defined in `ErrorCode.h`.
+
+For instance:
+
+```cpp
+#include <onnc/Support/ErrorCode.h>
+#include <onnc/Support/IOStream.h>
+#include <errno.h>
+
+using namespace onnc;
+
+int fd = open("/", O_WRONLY);
+SystemError err(errno); // get a copy from `errno`
+if (!err.isGood())
+  errs() << err; // permission deny
+```
+
+In this example, the `SystemError` type object `err` copies, translates, and stores the value of `errno`, a global variable representing error code given by the system.
+
+`errs()` is a reference to a OStream for standard error defined in [`<onnc/Support/IOStream.h>`](../../include/onnc/Support/IOStream.h).
+
+## Diagnostic
+
+The diagnostic system powers the capability for developers to find the problems. According to the severity of the information, there are seven levels defined in [`include <onnc/Diagnostic/MsgHandling.h>`](../../include/onnc/Diagnostic/MsgHandling.h)
+
+| Mode        | Handler           | Meaning  |
+| ------------- |:-------------:| -----|
+| normal | unreachable  | Errors that should never happen|
+| normal | fatal | Egregious errors that break the program|
+| normal |  error | Normal errors that the program keep going on until it cannot |
+| verbose | warning | warnings |
+| engineering verbose | debug | Information for debug mode |
+| engineering verbose | note | Less important information for debug mode |
+| noisy verbose | ignore | any verbose messages |
+
+The example shows how to use the handlers:
+
+```cpp
+#include <onnc/Diagnostic/MsgHandling.h>
+using namespace onnc;
+
+if (I got a serious problem)
+  fatal(fatal_open_folder) << "this folder" << 2;
+```
+
+The format is:
+
+```text
+handler(ErrorID) << mesg0 << mesg1 << ... << mesg9;
+```
+
+Note that there are **at most** `10` messages that can pass to a handler; otherwise ONNC shows the warning and breaks.
+
+ONNC supports the following types for messages:
+
+| Type | Definition |
+| ---- |------------|
+| bool/int32_t/int64_t | Support/DataTypes |
+| char*/std::string/onnc::StringRef | ADT/StringRef.h |
+| Path | Support/Path.h |
+| onnc::SystemError | Support/ErrorCode.h |
+
+### Registering a New Diagnostic Error
+
+To register a new diagnostic error, go to [`include/onnc/Diagnostics/DiagCommonKinds.inc`](../../include/onnc/Diagnostic/DiagCommonKinds.inc) and create a new `DIAG`.
+
+`DIAG`'s format is `DIAG(error name, error level, message)` where `message` is a string with placeholders `%N`, starts from `%0`, `%1` and so on, that receives inputs.
+
+For example, if a new error, `numerical_error`, is going to be registered. The line should be added to the end of the file `DiagCommonKinds.inc`:
+
+```inc
```

Comment:
should the style be `cpp` too?

Reply:
it's ok here, since Github can read the type.